File: gtci/18-bitwise-xor/cpp/3-two-single-numbers.cpp
#include <cassert>
#include <iostream>
#include <vector>

class Solution {
public:
  std::vector<int> findSingleNumbers(const std::vector<int>& nums) {
    int xorAll = 0;
    for (int num : nums)
      xorAll ^= num;

    // find the rightmost bit of xorAll
    // this is the rightmost bit where the two numbers differ
    const int rightmostBit = xorAll & -xorAll;

    // sort nums by the rightmost bit
    int xor1 = 0;
    int xor2 = 0;
    for (int num : nums) {
      if ((num & rightmostBit) != 0)
        xor1 ^= num;
      else
        xor2 ^= num;
    }

    return std::vector<int> {xor1, xor2};
  }
};

int main() {
    Solution s;
    
    auto result1 = s.findSingleNumbers({1, 4, 2, 1, 3, 2});
    assert((result1 == std::vector<int>{3, 4} || result1 == std::vector<int>{4, 3}));
    
    auto result2 = s.findSingleNumbers({1, 2, 3, 2});
    assert((result2 == std::vector<int>{1, 3} || result2 == std::vector<int>{3, 1}));
    
    auto result3 = s.findSingleNumbers({1, 2});
    assert((result3 == std::vector<int>{1, 2} || result3 == std::vector<int>{2, 1}));
    
    std::cout << "All test cases passed." << std::endl;
    return 0;
}

File: gtci/18-bitwise-xor/python3/3-two-single-numbers.py
class Solution:
    def findSingleNumbers(self, nums):
        xor_all = 0
        for num in nums:
            xor_all ^= num

        # get rightmost set bit in xor_all using bit trick
        # use this to find a bit where the two numbers differ
        rightmost_set_bit = xor_all & -xor_all

        # sort the two numbers into two groups using their differing rightmost bit
        xor_1 = 0
        xor_2 = 0
        for num in nums:
            if num & rightmost_set_bit != 0:
                # the bit is also set in num
                xor_1 ^= num
            else:
                xor_2 ^= num

        return [xor_1, xor_2]


if __name__ == "__main__":
    sol = Solution()

    assert sorted(sol.findSingleNumbers([1, 2, 1, 3, 2, 5])) == [3, 5]

    assert sorted(sol.findSingleNumbers([-1, 0])) == [-1, 0]

    assert sorted(sol.findSingleNumbers([0, 1])) == [0, 1]

    print("All test cases passed.")
